### PR TITLE
Cleanup superfluous let-bindings in 'removeUnsedExpr'

### LINKED ISSUE #1663: Unexpected data constructor in subject of `case` in netlist stage
While investigating https://github.com/clash-lang/clash-compiler/issues/1536 I found an issue where Clash errors on valid code. The error message is:

```
    Clash.Netlist(598): Not in normal form: Not a variable reference or primitive as subject of a case-statement:
    
    DataCon (SP "GHC.Maybe.Maybe" [("GHC.Maybe.Nothing",[]),("GHC.Maybe.Just",[SP "GHC.Maybe.Maybe" [("GHC.Maybe.Nothing",[]),("GHC.Maybe.Just",[Index 32768])]])]) (DC (SP "GHC.Maybe.Maybe" [("GHC.Maybe.Nothing",[]),("GHC.Maybe.Just",[SP "GHC.Maybe.Maybe" [("GHC.Maybe.Nothing",[]),("GHC.Maybe.Just",[Index 32768])]])],1)) [DataCon (SP "GHC.Maybe.Maybe" [("GHC.Maybe.Nothing",[]),("GHC.Maybe.Just",[Index 32768])]) (DC (SP "GHC.Maybe.Maybe" [("GHC.Maybe.Nothing",[]),("GHC.Maybe.Just",[Index 32768])],0)) []]
```

The core (obtained using `-fclash-debug DebugFinal`) confirms this is the case:

```
topEntity after normalization:

[..]
  = case <prefixName>"topEntity_ds1"
      Just @(Maybe (Index 32768)) (Nothing @(Index 32768)) of
      Nothing  ->
        <prefixName>"topEntity_ds1"
        Just @(Maybe (Index 32768)) (Nothing @(Index 32768))
      Just (ipv2 :: Maybe (Index 32768)) ->
        <prefixName>"topEntity_ds1"
        Just @(Maybe (Index 32768)) (Nothing @(Index 32768))
[..]
```

i.e.

```
topEntity after normalization:

[..]
  = case Just Nothing of
      Nothing  ->
        Just Nothing
      Just ipv2 ->
        Just Nothing
[..]
```

The netlist stage can only handle `case x of` where `x` is a variable reference and therefore crashes. Usually, this would have been cleaned up by `caseCon`, as the subject of the case is known. This constructs only appears _after_ executing `rmDeadCode` however, after which `caseCon` is never run again. What happens is:

```
let
  a = let foo = someComplexThing in Nothing
in
  case a of
    Nothing -> ...
    Just _ -> ...
```

If the (superfluous) `let` is not cleaned up, `bindConstantVar` will not inline it. In turn, `caseCon` will not fire.

Reproducer:

 * https://gist.github.com/martijnbastiaan/71b561250fec4e7e2b486cd0bdfea461

Solution:

 * Cleanup unused let-bindings in `removeUnusedExpr`

### PR DESCRIPTION
**Pull request by martijnbastiaan**
Fixes #1663

### CODE REVIEW

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 1]
```diff
@@ -953,6 +953,11 @@ removeUnusedExpr _ e@(collectArgsTicks -> (Data dc, [_,Right aTy,Right nTy,_,Lef
            in  changed v
       _ -> return e
 
+removeUnusedExpr _ e@(Letrec binds body) =
+  case removeUnusedBinders binds body of
```
Comment (christiaanb) [inline clash-lib/src/Clash/Normalize/Transformations.hs:957]: The issue with adding it to `removedUnusedExpr` is that the transformation is currently used in a topdown traversal, while you want to use `removeUnusedBinders` in a bottomup traversal after all the constants have been propagated and e.g. bodies of let-expressions stop refencing their binders. i.e. I would change: https://github.com/clash-lang/clash-compiler/blob/250aea43ce6111cd55cdb83e1ad30761024c9bb0/clash-lib/src/Clash/Normalize/Strategy.hs#L120

to

```haskell
inlineNR = bottomupR (apply "deadcode"  deadCode >-> apply "inlineNonRep" inlineNonRep)
```

Comment (martijnbastiaan) [reply clash-lib/src/Clash/Normalize/Transformations.hs]: Done! Albeit with a small twist :-)

File: clash-lib/src/Clash/Normalize/Strategy.hs [hunk 2]
```diff
@@ -117,7 +117,9 @@ constantPropagation =
     -- non-representable binder is inlined once, we can get rid of the recursive
     -- aspect using the case-of-known-constructor
     inlineNR :: NormRewrite
-    inlineNR = bottomupR (apply "inlineNonRep" inlineNonRep)
+    inlineNR =
+          bottomupR (apply "deadCode" deadCode)
+      >-! bottomupR (apply "inlineNonRep" inlineNonRep)
```
Comment (leonschoorl) [inline clash-lib/src/Clash/Normalize/Strategy.hs:122]: Wouldn't it be more efficient to combine the two transformations in one bottom up traversal?

Comment (martijnbastiaan) [reply clash-lib/src/Clash/Normalize/Strategy.hs:122]: We had some discussion on this on Slack, I'll update the commit message / PR message to explain the reasoning!

Comment (martijnbastiaan) [reply clash-lib/src/Clash/Normalize/Strategy.hs:134]: I've updated the comment right above the transformation!

**Pull request merged**